src/tests/suite/daos_nvme_recovery.c

```diff
@@ -94,28 +125,101 @@ nvme_recov_1(void **state)
 	}
 	ioreq_fini(&req);
 
-	print_message("Error injection to simulate device faulty.\n");
-	set_fail_loc(arg, rank, DAOS_NVME_FAULTY | DAOS_FAIL_ONCE);
+	/** Query test args to get total pool target count per node */
+	assert_true(arg->srv_ntgts > arg->srv_nnodes);
+	per_node_tgt_cnt = arg->srv_ntgts/arg->srv_nnodes;
 
-	/*
-	 * FIXME: Due to lack of infrastructures for checking each target
-	 *	  status, let's just wait for an arbitrary time and hope the
-	 *	  faulty reaction & rebuild is triggered.
+	/**
+	 * Verify initial states for all pool targets are UPIN by querying
+	 * the pool target info.
 	 */
+	for (i = 0; i < per_node_tgt_cnt; i++) {
+		rc = daos_pool_query_target(arg->pool.poh, i/*tgt*/,
+					    rank, &tgt_info, NULL /*ev*/);
+		assert_int_equal(rc, 0);
+		rc = strcmp(daos_target_state_enum_to_str(tgt_info.ta_state),
+			    "UPIN");
+		assert_int_equal(rc, 0);
+	}
+	print_message("All targets are in UPIN\n");
+
+	/** Inject error on random target index */
+	srand(time(NULL));
+	fail_loc_tgt = rand() % per_node_tgt_cnt;
+	print_message("Error injection on tgt %"PRIu64" to simulate device"
+		      " faulty.\n", fail_loc_tgt);
+	set_fail_loc(arg, rank, fail_loc_tgt, DAOS_NVME_FAULTY | DAOS_FAIL_ONCE);
+
+	/* Verify that the DAOS_NVME_FAULTY reaction got triggered. Target should
+	 * be in the DOWN state to trigger rebuild. If the target is still in
+	 * UPINT, attempt the faulty injection again with another target index.
+	 * The NVMe FAULTY device reaction might not have been triggered if the
+	 * target was not assigned to the device.
+	 */
+	while (fail_loc_tgt < per_node_tgt_cnt) {
+		rc = wait_and_verify_pool_tgt_state(arg->pool.poh, fail_loc_tgt,
+						    rank, "DOWN");
```

Comment:
btw, should we wait for "DOWN | DOWNOUT"?

Reply:
updated to accept multiple states separated by "|"